File: prototypes/python/import-project.py
import os
import json
import argparse
import shutil

# Utility Functions
def display_message(message='', prefix='', suffix='', verbose=True):
    if verbose:
        print(f"{prefix}{message}{suffix}")

def warning(message='', error_id=None, verbose=True):
    if verbose:
        display_message(message=message, prefix='Warning: ')
    if error_id is not None:
        raise SystemExit(error_id)

def newline(verbose=True):
    if verbose:
        print()

def import_config(config_file, verbose=True):
    if not config_file.endswith('.json'):
        config_file += '.json'
    if not os.path.exists(config_file):
        warning(f"Configuration file {config_file} is missing!", error_id=1, verbose=verbose)
    with open(config_file, 'r') as file:
        config = json.load(file)
    for key, value in config.items():
        os.environ[key] = str(value)
    if verbose:
        display_message(message=f"Imported configuration from {config_file}")

def apply_transformations(config_file, target_dir, verbose=True):
    with open(config_file, 'r') as file:
        config = json.load(file)

    for filename, replacements in config.items():
        target_file = os.path.join(target_dir, filename)
        if not os.path.exists(target_file):
            warning(f"Target file {target_file} does not exist!", verbose=verbose)
            continue

        # Read the content of the file
        with open(target_file, 'r') as f:
            content = f.read()

        # Apply replacements
        for placeholder, value in replacements.items():
            content = content.replace(f"${{{placeholder}}}", value)

        # Write the updated content back to the file
        with open(target_file, 'w') as f:
            f.write(content)

        if verbose:
            display_message(f"Applied transformations to {target_file}")

def setup_project_structure(source_dir, destination_dir, verbose=True):
    if not os.path.exists(source_dir):
        warning(f"Source directory {source_dir} does not exist!", error_id=2, verbose=verbose)
    if os.path.exists(destination_dir):
        shutil.rmtree(destination_dir)
    shutil.copytree(source_dir, destination_dir)
    if verbose:
        display_message(f"Copied project structure from {source_dir} to {destination_dir}")

def parse_arguments():
    parser = argparse.ArgumentParser(description='Process some parameters.')
    parser.add_argument('--config', type=str, required=True, help='Configuration file')
    parser.add_argument('--target-dir', type=str, required=True, help='Target directory for transformations')
    parser.add_argument('--source-dir', type=str, required=True, help='Source directory for project structure')
    parser.add_argument('--destination-dir', type=str, required=True, help='Destination directory for project setup')
    parser.add_argument('--verbose', action='store_true', help='Enable verbose output')
    return parser.parse_args()

def main():
    args = parse_arguments()

    # Import environment variables from the configuration file
    import_config(args.config, verbose=args.verbose)
    newline(verbose=args.verbose)

    # Set up project structure
    setup_project_structure(args.source_dir, args.destination_dir, verbose=args.verbose)
    newline(verbose=args.verbose)

    # Apply transformations to files in the target directory
    apply_transformations(args.config, args.target_dir, verbose=args.verbose)
    newline(verbose=args.verbose)

if __name__ == '__main__':
    main()

File: prototypes/ruby/import-project.rb
# Ruby Port of the Shell/Bash Script
def display_message(message: '', prefix: '', suffix: '', verbose: true)
  puts "#{prefix}#{message}#{suffix}" if verbose
end

def warning(message: '', error_id: nil, verbose: true)
  display_message(message: message, prefix: 'Warning: ', verbose: verbose)
  exit(error_id) if error_id
end

def newline(verbose: true)
  puts if verbose
end

def import_config(config_file, verbose: true)
  config_file += '.json' unless config_file.end_with?('.json')
  unless File.exist?(config_file)
    warning(message: "Configuration file #{config_file} is missing!", error_id: 1, verbose: verbose)
  end

  config = JSON.parse(File.read(config_file))
  config.each { |key, value| ENV[key] = value.to_s }

  display_message(message: "Imported configuration from #{config_file}", verbose: verbose)
end

def apply_transformations(config_file, target_dir, verbose: true)
  config = JSON.parse(File.read(config_file))

  config.each do |filename, replacements|
    target_file = File.join(target_dir, filename)
    unless File.exist?(target_file)
      warning(message: "Target file #{target_file} does not exist!", verbose: verbose)
      next
    end

    # Read the content of the file
    content = File.read(target_file)

    # Apply replacements
    replacements.each do |placeholder, value|
      content.gsub!("${#{placeholder}}", value)
    end

    # Write the updated content back to the file
    File.write(target_file, content)

    display_message(message: "Applied transformations to #{target_file}", verbose: verbose)
  end
end

def setup_project_structure(source_dir, destination_dir, verbose: true)
  unless Dir.exist?(source_dir)
    warning(message: "Source directory #{source_dir} does not exist!", error_id: 2, verbose: verbose)
  end

  if Dir.exist?(destination_dir)
    FileUtils.rm_rf(destination_dir)
  end

  FileUtils.cp_r(source_dir, destination_dir)
  display_message(message: "Copied project structure from #{source_dir} to #{destination_dir}", verbose: verbose)
end

def parse_arguments
  require 'optparse'

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: script.rb [options]"

    opts.on("--config CONFIG", "Configuration file") { |v| options[:config] = v }
    opts.on("--target-dir DIR", "Target directory for transformations") { |v| options[:target_dir] = v }
    opts.on("--source-dir DIR", "Source directory for project structure") { |v| options[:source_dir] = v }
    opts.on("--destination-dir DIR", "Destination directory for project setup") { |v| options[:destination_dir] = v }
    opts.on("--verbose", "Enable verbose output") { options[:verbose] = true }
  end.parse!

  options
end

def main
  args = parse_arguments

  # Import environment variables from the configuration file
  import_config(args[:config], verbose: args[:verbose])
  newline(verbose: args[:verbose])

  # Set up project structure
  setup_project_structure(args[:source_dir], args[:destination_dir], verbose: args[:verbose])
  newline(verbose: args[:verbose])

  # Apply transformations to files in the target directory
  apply_transformations(args[:config], args[:target_dir], verbose: args[:verbose])
  newline(verbose: args[:verbose])
end

if __FILE__ == $0
  require 'json'
  require 'fileutils'
  main
end
